pgrx/src/pgbox.rs

```diff
@@ -368,15 +369,64 @@ impl<T, AllocatedBy: WhoAllocated> PgBox<T, AllocatedBy> {
     }
 }
 
+impl<T> Clone for PgBox<T, AllocatedByPostgres>
+where
+    T: Clone,
+{
+    /// Copies the wrapped `T` into [`PgMemoryContexts::CurrentMemoryContext`].
+    fn clone(&self) -> Self {
+        unsafe {
+            let copy = PgMemoryContexts::CurrentMemoryContext
+                .copy_ptr_into(self.as_ptr(), std::mem::size_of::<T>());
+
+            PgBox::from_pg(copy)
+        }
+    }
+}
+
+impl<T, AllocatedBy: WhoAllocated> Eq for PgBox<T, AllocatedBy> where T: Eq {}
+impl<T, AllocatedBy: WhoAllocated> PartialEq for PgBox<T, AllocatedBy>
+where
+    T: PartialEq,
+{
+    fn eq(&self, other: &Self) -> bool {
+        self.as_ref() == other.as_ref()
+    }
+}
+
+impl<T, AllocatedBy: WhoAllocated> Debug for PgBox<T, AllocatedBy>
+where
+    T: Debug,
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
+        write!(f, "{:?}", self.as_ref())
+    }
+}
+
+impl<T, AllocatedBy: WhoAllocated> Display for PgBox<T, AllocatedBy>
+where
+    T: Display,
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
+        write!(f, "{}", self.as_ref())
+    }
+}
+
+impl<T, AllocatedBy: WhoAllocated> AsRef<T> for PgBox<T, AllocatedBy> {
+    fn as_ref(&self) -> &T {
+        match self.ptr.as_ref() {
+            Some(ptr) => unsafe { ptr.as_ref() },
+            None => panic!("Attempt to dereference null pointer during `AsRef::as_ref()` of PgBox"),
+        }
+    }
+}
```

Comment:
feeling nervous about the soundness of this one, even with the null check, since it's easy to create a PgBox from a potentially-dubious pointer.

Reply:
I can not impl `AsRef` and instead add a `unsafe fn as_ref(self) -> &T` function to `PgBox` that does this?  Maybe have it return `Option<&T>`?

I did this only to make that roundtrip test macro a little easier to write for PgBox'd things.

Reply:
I had mostly decided there was probably something cursed about a new impl to set this weird cascade of failures off, and started looking for something dubious, but I think this may be less of a problem than I thought.

Reply:
I've decided this is no more dubious than PgBox already was.